### PR TITLE
make basic debugging possible

### PR DESCRIPTION
**Pull request by krux02**
the debug procs once in the past did not need a config object. Now they don't need it anymore. And by default it is set to nil. This doesn't work as well as if it is set, but it won't crash anymore on nil. Also the order of the arguments is reversed so that the old way to use the debug proc will just work.

I also added debug information to the debug build of the compiler. 

With these changes basic debugging of the compiler is now possible. Still a long road ahead to make debugging a better experience. But these steps are absolutely necessary.

### CODE REVIEW

File: koch.nim [hunk 1]
```diff
@@ -470,7 +470,7 @@ proc temp(args: string) =
   # commit.
   let (bootArgs, programArgs) = splitArgs(args)
   let nimexec = findNim()
-  exec(nimexec & " c -d:debug " & bootArgs & " compiler" / "nim", 125)
+  exec(nimexec & " c --debugger:native " & bootArgs & " compiler" / "nim", 125)
```
Comment (Araq) [inline koch.nim]: This does not replace ``-d:debug``!

Comment (krux02) [reply koch.nim]: Ok I added `-d:debug` again. Is this ok now?

**Comment by zah** [comment]
I have a slightly more powerful fix here:
https://github.com/nim-lang/Nim/blob/e129466910b6efa730f8d5d9232efbce6dae46f0/compiler/astalgo.nim#L34

Waiting on the typedesc reforms pull requests to be finally merged:
https://github.com/nim-lang/Nim/pull/7681/

**Comment by zah** [comment]
I would appreciate it if my pull request is merged and this is rebased on top of it. The few remaining issues in the pull request can be addressed later.

**Comment by krux02** [comment]
@zah Not sure if I like your "more powerful fix". I don't like it at all when templates use identifiers from the environment at all. It makes it very hard to read the code. And the fallback to error is exactly what I do want to prevent. "debug" should not error at all. It should do whatever it can do and fall back to no or incomplete output when it can't do a better job, but it may not raise an error.

**Comment by zah** [comment]
`debug` triggers a compile-time error in my version, but as I said it can fallback to your `nil` handling code.

Otherwise, my fix is better simply because it produces better output for the same effort from the developer. These are internal helpers used only during development and the usual "good design" criteria don't apply to them.

**Comment by krux02** [comment]
@zah debug simply has different requirements for good design. It is tool you use frequently to dump information about nodes on the console and then remove it again when you got what you were looking for.  So it should be short easy to remember, and easy to get the parameters right. You don't want to fiddle for a long time with compilation errors just to get all the parameters right.

But I agree your version combined with my fallback on nil could be the best implementation for debug. 

By the way calling debug is also very useful, when you call it from gdb.  Dumping arbitrary nodes when exploring the stack frame is indeed very useful. This way you don't need to recompile the compiler with debug annotations, you just invoke debug manually as soon as you hit a breakpoint. Here it would be useful, if there is a version of debug that does not rely on any template magic or static overloading at all.

File: koch.nim [hunk 2]
```diff
@@ -470,7 +470,7 @@ proc temp(args: string) =
   # commit.
   let (bootArgs, programArgs) = splitArgs(args)
   let nimexec = findNim()
-  exec(nimexec & " c -d:debug " & bootArgs & " compiler" / "nim", 125)
+  exec(nimexec & " c -d:debug --debugger:native " & bootArgs & " compiler" / "nim", 125)
```
Comment (Varriount) [inline koch.nim]: Why was this added?

Comment (krux02) [reply koch.nim]: so that the nim_temp build has debug information. Debug information is necessary to use debuggers like gdb. Well to be precise one can debug without debug information, but then then you debug in disassembled binary instead of seeing the actual source.

Comment (krux02) [reply koch.nim]: Just try it, and run ``gdb nim_temp``.  There are two possible outputs:

```
Reading symbols from nim_temp...done.
```

```
Reading symbols from nim_temp...(no debugging symbols found)...done.
```
